server/pkg/docker/build.go

```diff
@@ -0,0 +1,156 @@
+}
+
+func BuildReleaseArtifacts(ctx context.Context, opts BuildReleaseArtifactsOpts, logger hclog.Logger) (error, func() error) {
+	serviceDirInContext := ".trdl"
+	serviceDockerfilePathInContext := path.Join(serviceDirInContext, "Dockerfile")
+	serviceLabels := map[string]string{
+		"vault-trdl-release-uuid": uuid.NewV4().String(),
+	}
+
+	contextBuf := buffer.New(64 * 1024 * 1024)
+	contextReader, contextWriter := nio.Pipe(contextBuf)
+
+	secrets, err := secrets.GetSecrets(ctx, opts.Storage)
+	if err != nil {
+		return fmt.Errorf("unable to get build secrets: %w", err), nil
+	}
+
+	go func() {
+		if err := func() error {
+			tw := tar.NewWriter(contextWriter)
+
+			logboek.Context(ctx).Default().LogF("Adding git worktree files to the build context\n")
+			logger.Debug("Adding git worktree files to the build context")
+
+			if err := trdlGit.AddWorktreeFilesToTar(tw, opts.GitRepo); err != nil {
+				return fmt.Errorf("unable to add git worktree files to tar: %w", err)
+			}
+
+			dockerfileOpts := DockerfileOpts{
+				Labels:  serviceLabels,
+				Secrets: secrets,
+			}
+			if err := GenerateAndAddDockerfileToTar(tw, serviceDockerfilePathInContext, opts.FromImage, opts.RunCommands, dockerfileOpts); err != nil {
+				return fmt.Errorf("unable to add service dockerfile to tar: %w", err)
+			}
+
+			if err := tw.Close(); err != nil {
+				return fmt.Errorf("unable to close tar writer: %w", err)
+			}
+
+			return nil
+		}(); err != nil {
+			if closeErr := contextWriter.CloseWithError(err); closeErr != nil {
+				panic(closeErr)
+			}
+			return
+		}
+
+		if err := contextWriter.Close(); err != nil {
+			panic(err)
+		}
+	}()
+
+	logboek.Context(ctx).Default().LogF("Building docker image with artifacts\n")
+	logger.Debug("Building docker image with artifacts")
+
+	args, err := setCliArgs(serviceDockerfilePathInContext, secrets)
+	if err != nil {
+		return fmt.Errorf("unable to set cli args: %w", err), nil
+	}
+
+	cli, err := newDockerCli(defaultCliOptions(ctx))
+	if err != nil {
+		return fmt.Errorf("error creating docker cli: %w", err), nil
+	}
+
+	if err := CliBuild(cli, contextReader, opts.TarWriter, args...); err != nil {
+		return err, nil
+	}
+
+	cleanupFunc := func() error {
+		cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			return fmt.Errorf("unable to create docker client: %w", err)
+		}
+		return RemoveImagesByLabels(ctx, cli, serviceLabels)
+	}
+
+	return nil, cleanupFunc
+}
+
+func CliBuild(c command.Cli, contextReader *nio.PipeReader, tarWriter *nio.PipeWriter, args ...string) error {
+	var finalArgs []string
+	var cmd *cobra.Command
+
+	c.SetIn(streams.NewIn(contextReader))
+
+	cmd = NewBuildxCommand(c)
+	finalArgs = append([]string{"build"}, args...)
+
+	cmd = prepareCliCmd(cmd, finalArgs...)
+
+	reader, writer, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	origStdOut := os.Stdout
+	os.Stdout = writer
+	defer func() { os.Stdout = origStdOut }()
```

Comment:
This approach might cause issues in real-world scenarios since os.Stdout could be used by other tasks or threads. Additional investigation is needed to assess potential side effects.